### PR TITLE
[WIP] Attempt a Webpack v4 support

### PR DESCRIPTION
**Pull request by astorije**

### CODE REVIEW

**Comment by astorije** [comment]
Alright, this is not working too well, I could use some help 😅

**Comment by zinserjan** [comment]
Thanks for working on this :sunglasses:


Watch stuff seems to fail. Looks like webpack moved their watch compiler into it's own file and it's no longer available on the Compiler as Watching property.

Try to replace 

**mocha-webpack/src/webpack/compiler/createWatchCompiler.js**
```js
const createWatcher = () => new compiler.constructor.Watching(compiler, watchOptions, noop);
```

with
```js
import Watching from "webpack/lib/Watching");
...

const createWatcher = () => new Watching(compiler, watchOptions, noop);
```

All removed features like `module.loaders` should be also removed from mocha-webpack.

I think we should also drop support for older mocha versions, cause webpack@4 support is already a breaking change and needs to be released as 2.0.

**Comment by astorije** [comment]
Hey @zinserjan, thanks for your answer!
Unfortunately, I have to admit that I have no idea what I'm doing at this point, and it's definitely not working.
If anyone wants to help, I'm happy to give write access to my branch. @zinserjan, you are able to write to my branch as part of this PR.

I would love to be able to finish this, but clearly I'm not :( Sorry!

> I think we should also drop support for older mocha versions, cause webpack@4 support is already a breaking change and needs to be released as 2.0.

It would be nice to keep at least support for Mocha v4, because it's been released not that long ago, but obviously, your call. As long as we can make Mocha v5 + Webpack v4 work, I'm happy :D

**Comment by zinserjan** [comment]
There were a few problems:
- Errors happened in prepareMocha were never catched.
- Tests used only stdout to analyse the results. Added a way to merge stdout and stderr
- Skipped watch tests temporary, we can start with watch mode when single run works

=> we see finally whats wrong

File: src/cli/requireWebpackConfig.js [hunk 1]
```diff
@@ -104,6 +104,10 @@ export default function requireWebpackConfig(webpackConfig, required, env) {
     config = config(env);
   }
 
+  if (!config.mode) {
+    config.mode = mode;
```
Comment (zinserjan) [inline src/cli/requireWebpackConfig.js]: I think we should overwrite mode in webpack config when mode is supplied via cli options

**Comment by zinserjan** [comment]
Tests are green now :grin:
But eslint fails on some files with `no-undef` failure and I have no idea why..

**Comment by astorije** [comment]
omg this is so so so awesome of you, I can't believe you found the time to actually fix up my PR ❤️ 

I googled the internets, and it seemed like this `no-undef` error was caused by a combination of old `eslint`/`babel-eslint` versions. Bumping those 2 dependencies magically fixed the `no-undef`s (but I had to fix other errors, which is good as they seemed legit).
I'm guessing the reason why this appeared on this branch and not master is because of some sub-sub-dependency version that bumping Webpack picked up along the way.

Note that I have not changed anything about the linting configuration, just bumped major versions blindly, so there might be new/updated rules that you want to add, I just did the minimum required to make the checks pass.
Similarly, with the current combination of versions, you now get the following warnings at install time:
```
warning " > eslint-config-airbnb-base@4.0.2" has incorrect peer dependency "eslint@^3.0.1".
warning " > eslint-plugin-import@1.16.0" has incorrect peer dependency "eslint@2.x - 3.x".
```
I didn't bump those dependencies, might be worth doing that after this PR is all set and is merged, but I also figured that would come with changes to the config and you'd want to control that.

Thanks again for doing that!

**Comment by zinserjan** [comment]
Yeah awesome, thanks!

I think I would like to replace flow with typescript and the eslint stuff with prettier instead of configuring all the time :laughing:

**Comment by zinserjan** [comment]
Good question, but I think requiresArg means just when you provide this specific option you have to provide also a value.. . I'll verify this.

**Comment by zinserjan** [comment]
Code-splitting tests runs without `--mode` and just use the config.

**Comment by zinserjan** [comment]
@astorije I think it's time to merge this in or is there anything left?

**Comment by astorije** [comment]
@zinserjan, I am getting the following exception when running our test suite against https://github.com/thelounge/thelounge/pull/2117 + this branch (loaded with `"mocha-webpack": "file:../mocha-webpack"` after running `yarn build`):

```
RUNTIME EXCEPTION  Exception occurred while loading your tests

TypeError: Cannot read property 'call' of undefined
   at __webpack_require__ ([...]/.tmp/mocha-webpack/1521390715327/webpack:/webpack/bootstrap:22:1)
   at __dirname../client/js/constants.js.path ([...]/.tmp/mocha-webpack/1521390715327/webpack:/webpack/bootstrap:74:1)
   at Object.<anonymous> ([...]/.tmp/mocha-webpack/1521390715327/bundle.js:78:10)
   at Module._compile (module.js:643:30)
   at Object._module2.default._extensions.(anonymous function) [as .js] ([...]/node_modules/mocha-webpack/lib/util/registerRequireHook.js:147:12)
   at Module.load (module.js:556:32)
   at tryModuleLoad (module.js:499:12)
   at Function.Module._load (module.js:491:3)
   at Module.require (module.js:587:17)
   at require (internal/module.js:11:18)
   at [...]/node_modules/mocha-webpack/node_modules/mocha/lib/mocha.js:231:27
```

Not sure if that's an issue with this branch or my setup.

**Comment by zinserjan** [comment]
@astorije sounds like webpack/webpack#6727

**Comment by astorije** [comment]
Oh, yes, I forgot we removed [this](https://github.com/thelounge/thelounge/pull/2117/files#diff-05b8e9c5eeb66ddb602cd5c51aa96f6eL22)!

I then confirm that this PR works fine against the repo I linked!

🚢 :shipit:❗️

**Comment by astorije** [comment]
🎉

**Comment by ctaylo21** [comment]
Great work guys!

**Pull request merged**